### PR TITLE
Entirely Reworks Ambience

### PR DESCRIPTION
**Pull request by qwertyquerty**
# About The Pull Request

Completely refactors ambient noises into a subsystem called ambience! This clears up a lot of the horrid code that was old ambience, which was strung together by shoestrings and elmers glue sticks. Here's how it works now:

### BUZZ

- The background whirrrrrrrrrrrrrrrrrrr of the station, plays on loop
- Completely IC, so the user needs to be able to hear to hear it, if at any point you shouldn't be able to hear it, it will immediately stop
- Being able to hear entails not being deaf and not being in a very low pressure area
- Can only specify one per area, but can have different areas play different buzzes (aka different station whirr vs shuttle whirr)
- The pre-existing station buzz has been refactored into this

### EFFECTS

- Play at random (minimum time between them being 60 seconds)
- The random sound effects that play on the station, like the airport one
- Usually pretty short
- Also IC, so you need to be able to hear it
- However, once these start they will play to completion instead of stopping once you can't hear it
- There's tons of these
- You can have multiple per area, and it will pick one at random
- All of the previous random ambient sound effects have been refactored into this

### MUSIC

- Completely OOC, you don't need to be able to hear in game to hear these
- Play to completion once started
- Have a 1 in 80 chance of playing every second if you are in an area that has them
- Will play until completion, won't let another one play until one is done
- Perfect for eerie or adventurous space music (all the space ambience has been refactored into this category, as you wouldn't be able to hear it in the other two categories because there's no pressure in space!!)
- You can specify multiple per area, will pick at random
- Plays a bit louder than buzz or effects

## Why It's Good For The Game

Bad code is bad, and we like improving the codebase here at beestation r tm. Also ambient effects improve muh immersion.

## Changelog
:cl:
refactor: Refactors ambient sounds completely to be less bad code wise and better IC and immersion wise
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
Holy shit, ambience system that isnt shit, nice

File: code/controllers/subsystem/ambience.dm [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+	else if (M.client.ambient_buzz_playing) // If it's playing, and it shouldn't be, stop it
+		M.stop_sound_channel(CHANNEL_AMBIENT_BUZZ)
+		M.client.ambient_buzz_playing = null
+
+
+/datum/controller/subsystem/ambience/proc/update_music(mob/M) // Background music, the more OOC ambience, like eerie space music
+	var/area/A = get_area(M)
+
+	if (A.ambient_music && (M.client.prefs.toggles & SOUND_AMBIENCE) && (prob(1) && prob(25)) && !M.client.channel_in_use(CHANNEL_AMBIENT_MUSIC)) // 1/80 chance to play every second, only play while another one is not playing
```
Comment (PowerfulBacon) [inline code/controllers/subsystem/ambience.dm]: if(prob(1) && prob(25))

**Comment by PowerfulBacon** [comment]
The music and effects dont really need to update 5 times per second, could probably optimise it a bit more if you only update music every 5th tick or something

**Comment by Kmc2000** [comment]
Could you add something similar to our looping_ambience field on /area/? that'd be neat to avoid conflicts!

(Looping ambience plays the ambience specified on repeat until you leave the room)

**Comment by qwertyquerty** [comment]
@Kmc2000 isn't that just buzz

File: code/controllers/subsystem/ambience.dm (hunk 1 above)
Comment (qwertyquerty) [reply code/controllers/subsystem/ambience.dm:59]: > The music and effects dont really need to update 5 times per second, could probably optimise it a bit more if you only update music every 5th tick or something

@PowerfulBacon if they don't update fast you don't get the effect of stuff like buzz stopping and starting immediately when you walk out into space and back in

This subsystem is extremely inexpensive and I think it's fine

**Comment by qwertyquerty** [comment]
> The music and effects dont really need to update 5 times per second, could probably optimise it a bit more if you only update music every 5th tick or something

@PowerfulBacon if they don't update fast you don't get the effect of stuff like buzz stopping and starting immediately when you walk out into space and back in

This subsystem is extremely inexpensive and I think it's fine

File: code/controllers/subsystem/ambience.dm (hunk 1 above)
Comment (qwertyquerty) [reply code/controllers/subsystem/ambience.dm]: Yes a 1 in 400 chance

**Comment by qwertyquerty** [comment]
> The music and effects dont really need to update 5 times per second, could probably optimise it a bit more if you only update music every 5th tick or something

Oh you just said music and effects, yeah you right I'll make that change later

File: code/controllers/subsystem/ambience.dm (hunk 1 above)
Comment (qwertyquerty) [reply code/controllers/subsystem/ambience.dm]: fixed

**Comment by qwertyquerty** [comment]
@PowerfulBacon music and buzz now update every 5th fire

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**